src/Assistant.ts

```diff
@@ -0,0 +1,376 @@
+import {
+  AssistantThreadsSetStatusResponse,
+  AssistantThreadsSetSuggestedPromptsResponse,
+  AssistantThreadsSetTitleResponse,
+  ChatPostMessageArguments,
+} from '@slack/web-api';
+import processMiddleware from './middleware/process';
+import {
+  AllMiddlewareArgs,
+  AnyMiddlewareArgs,
+  Middleware,
+  SayFn,
+  SlackEventMiddlewareArgs,
+} from './types';
+import { AssistantInitializationError, AssistantMissingPropertyError } from './errors';
+import {
+  AssistantThreadContext,
+  AssistantThreadContextStore,
+  DefaultThreadContextStore,
+  GetThreadContextFn,
+  SaveThreadContextFn,
+} from './AssistantThreadContextStore';
+
+/**
+ * Configuration object used to instantiate the Assistant
+ */
+export interface AssistantConfig {
+  threadContextStore?: AssistantThreadContextStore;
+  threadStarted: AssistantThreadStartedMiddleware | AssistantThreadStartedMiddleware[];
+  threadContextChanged?: AssistantThreadContextChangedMiddleware | AssistantThreadContextChangedMiddleware[];
+  userMessage: AssistantUserMessageMiddleware | AssistantUserMessageMiddleware[];
+}
+
+/**
+ * Callback utilities
+ */
+interface AssistantUtilityArgs {
+  getThreadContext: GetThreadContextFn;
+  saveThreadContext: SaveThreadContextFn;
+  say: SayFn;
+  setStatus: SetStatusFn;
+  setSuggestedPrompts: SetSuggestedPromptsFn;
+  setTitle: SetTitleFn;
+}
+
+export interface SetStatusFn {
+  (status: string): Promise<AssistantThreadsSetStatusResponse>;
+}
+
+export interface SetSuggestedPromptsFn {
+  (params: SetSuggestedPromptsArguments): Promise<AssistantThreadsSetSuggestedPromptsResponse>;
+}
+
+export interface SetSuggestedPromptsArguments {
+  prompts: [AssistantPrompt, ...AssistantPrompt[]];
+}
+
+interface AssistantPrompt {
+  title: string;
+  message: string;
+}
+
+export interface SetTitleFn {
+  (title: string): Promise<AssistantThreadsSetTitleResponse>;
+}
+
+/**
+ * Middleware
+ */
+export type AssistantThreadStartedMiddleware = Middleware<AssistantThreadStartedMiddlewareArgs>;
+export type AssistantThreadContextChangedMiddleware = Middleware<AssistantThreadContextChangedMiddlewareArgs>;
+export type AssistantUserMessageMiddleware = Middleware<AssistantUserMessageMiddlewareArgs>;
+
+export type AssistantMiddleware =
+  | AssistantThreadStartedMiddleware[]
+  | AssistantThreadContextChangedMiddleware[]
+  | AssistantUserMessageMiddleware[];
+
+export type AssistantMiddlewareArgs =
+  | AssistantThreadStartedMiddlewareArgs
+  | AssistantThreadContextChangedMiddlewareArgs
+  | AssistantUserMessageMiddlewareArgs;
+
+export interface AssistantThreadStartedMiddlewareArgs extends
+  Omit<SlackEventMiddlewareArgs<'assistant_thread_started'>, 'say'>, AssistantUtilityArgs {}
```

Comment:
Do I have it right that the reason we are omitting `say` here is that the utility `WhenEventHasChannelContext` used in `SlackEventMiddlewareArgs` gets in the way for these agent events? I guess the assumption this utility type operates on (always tack on a `say` if the event has a channel context) is not right - or is it? I'm poking this 'cause it is weird to omit `say`, but then add it right back via `AssistantUtilityArgs` - our abstractions are not lining up.

If so, let's drop a TODO in here to resolve the leaky abstraction for `say` in assistant middleware args in v5.

Reply:
IIRC, I had to omit it here because when the event comes in, there is no `say` associated with the payload type's args. It only becomes relevant, as you've pointed out, when we add it back later on. I'll revisit this and see if anything has changed since I wrote these lines, and, if not, will add a TODO.